### PR TITLE
 Resource state visualization

### PR DESCRIPTION
**Pull request by d1ssk**
Before submitting, please check the following:
- Make sure you have tests for the new code and that test passes (run `pytest`)
- format added code and tests by `black -l 120 <filename>`
- If applicable, add a line to the [unreleased] part of CHANGELOG.md, following [keep-a-changelog](https://keepachangelog.com/en/1.0.0/).
 
Then, please fill in below:

**Context (if applicable):**

**Description of the change:**
Implement a `GraphVisualizer` class within visualization.py and add a `view_graph()` method to the `Pattern` class, which invokes methods from the `GraphVisualizer` class.
- The graph states are visualized with flow structure (or gflow structure, to be implemented). 
- Input and output nodes are distinguished by edge color and the color of the node faces respectively.
- (Optional) Nodes measured in the Pauli bases are colored blue.
- (Optional) Indexes of the local Clifford operator are displayed adjacent to the nodes.

**Related issue:**
#47 

also see that checks (github actions) pass.
If lint check keeps failing, try installing black==22.8.0 as behavior seems to vary across versions.

### CODE REVIEW

**Review by pafloxy — COMMENTED** [review]
Thanks for the PR. I have tried out the plotting subroutines you have introduced and they work really smooth and the plots are visually appleasing.

I had one suggestion about it though, in `visualize_w_flow` for building the intial layout perhaps it might be helpful to check the `nx.multipartite_layout()` function which can generate a layered structure as well given the layer information is passed on to the graph as a `node_attribrute`. 

These method would in particular be helpful for the case of plotting resource states with gflow as the method used here sets the y-axis coordinates based on the sequence of flow edges connecting an input node to its corresponding output node, which might not be possible for the case where the gflow maps a node to multiplle nodes. 

Let me know what you think :)

File: graphix/vizualization.py [hunk 1]
```diff
@@ -0,0 +1,221 @@
...
+            Filename of the saved plot.
+        """
+        if figsize is None:
+            width = (max(l_k.values()) + 1) * 0.8
+            height = len(self.v_in)
+            figsize = (width, height)
+        plt.figure(figsize=figsize)
+        pos = nx.spring_layout(self.g)  # Initial layout for the nodes
+
+        n = len(self.v_in)
```
Comment (pafloxy) [inline graphix/vizualization.py]: The part I commented about in my [review](https://github.com/TeamGraphix/graphix/pull/78#pullrequestreview-1588217396)

**Comment by d1ssk** [comment]
I was just wondering how to set the position of nodes in the case of gflow and your suggestion looks great!
I will use `nx.multipartite_layout()` for visualization of both flow and gflow.

**Comment by pafloxy** [comment]
Sure! I'll be happy to discuss if you want.  

However in my experience, time to time the `nx.multipartite_layout` does behave weirdly in placing the graph nodes which I think we will need to take care of it somehow in the post-processing. 

For eg, see 
image
the nodes 7, 28, 21 are unnecessarily pushed out of line. 

###### The figure is from a code I wrote for the visualisations but ofcourse it isn't even closely sophisticated compared to your implementation :sweat_smile:

**Comment by d1ssk** [comment]
Thank you for the valuable information.  I would like to reproduce this example to consider a countermeasure, could you give me the code that generates this graph?

**Comment by pafloxy** [comment]
Sure, you will find a fucntion called `pretty_plot_pattern()` defined in this [notebook](https://github.com/pafloxy/graphix_pafloxy/blob/47-resource-state-visualization/scratchnb_0.ipynb) It has the required implementation. Though it isn't really a good peice of programming.

File: graphix/pattern.py [hunk 2]
```diff
@@ -888,6 +889,36 @@ def get_meas_plane(self):
...
+        local_clifford : dict
+            local clifford index of the each node.
+        """
+        local_clifford = {}
+        for cmd in self.seq:
+            if cmd[0] == "C":
+                local_clifford[cmd[1]] = cmd[2]
+            elif cmd[0] == "M" and len(cmd) == 7:
+                local_clifford[cmd[1]] = cmd[6]
+        return local_clifford
```
Comment (shinich1) [inline graphix/pattern.py:920]: isn't this same as get_vops?

File: graphix/pattern.py [hunk 3]
```diff
@@ -1212,6 +1243,40 @@ def perform_pauli_measurements(self, leave_input=False):
...
+        if pauli_indicator:
+            angles = self.get_angles()
+        else:
+            angles = None
+        if local_clifford_indicator:
+            local_clifford = self.get_local_clifford()
+        else:
+            local_clifford = None
+        vis.visualize(figsize=figsize, angles=angles, local_clifford=local_clifford, save=save, filename=filename)
+
```
Comment (shinich1) [inline graphix/pattern.py:1279]: this should be named `visualize_flow` instead, perhaps?

**Review by shinich1 — APPROVED** [review]
once you worked on the comments above and merged in updates of master branch, please squash and merge.

File: graphix/pattern.py (hunk 3 above)
Comment (d1ssk) [reply graphix/pattern.py]: This method also visualize gflow graph or no flow/gflow graph, so a name with a broader meaning is preferred.
I set the name `draw_graph()`

**Pull request merged**